lib/ansible/modules/clustering/oc.py

```diff
@@ -0,0 +1,471 @@
+    to be present.  In the case of a deletion, this is the response from the
+    delete request.
+  returned: success
+  type: string
+url:
+  description: The URL to the requested resource.
+  returned: success
+  type: string
+method:
+  description: The HTTP method that was used to take action upon the resource
+  returned: success
+  type: string
+...
+'''
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.pycompat24 import get_exception
+from ansible.module_utils import urls
+from ansible.module_utils.six.moves.urllib.parse import urlencode
+import copy
+
+class ApiEndpoint(object):
+    def __init__(self, host, port, api, version):
+        self.host = host
+        self.port = port
+        self.api = api
+        self.version = version
+
+    def __str__(self):
+        url = "https://"
+        url += self.host
+        url += ":"
+        url += str(self.port)
+        url += "/"
+        url += self.api
+        url += "/"
+        url += self.version
+        return url
+
+class ResourceEndpoint(ApiEndpoint):
+    def __init__(self, name, namespaced, api_endpoint):
+        super(self.__class__, self).__init__(api_endpoint.host,
+                                             api_endpoint.port,
+                                             api_endpoint.api,
+                                             api_endpoint.version)
+        self.name = name
+        self.namespaced = namespaced
+
+class NamedResource(object):
+    def __init__(self, module, definition, resource_endpoint):
+        self.module = module
+        self.set_definition(definition)
+        self.resource_endpoint = resource_endpoint
+
+    def name(self):
+        if 'name' in self.definition['metadata'].keys():
+            return self.definition['metadata']['name']
+        return None
+
+    def namespace(self):
+        if 'namespace' in self.definition['metadata'].keys():
+            return self.definition['metadata']['namespace']
+        return None
+
+    def set_definition(self, definition):
+        if isinstance(definition, str):
+            self.definition = self.module.from_json(response.read())
+        else:
+            self.definition = definition
+
+    def url(self, create=False):
+        url = str(self.resource_endpoint)
+        url += '/'
+        if self.resource_endpoint.namespaced:
+            url += 'namespaces/'
+            url += self.namespace()
+            url += '/'
+        url += self.resource_endpoint.name
+        if not create:
+            url += '/'
+            url += self.name()
+        return url
+
+    def __dict__(self):
+        return self.definition
+
+    def __str__(self):
+        return self.module.jsonify(self.definition)
+
+
+class OC(object):
+    def __init__(self, module, token, host, port,
+                 apis=['api', 'oapi']):
+        self.apis = apis
+        self.version = 'v1'
+        self.token = token
```

Comment:
For ease of use, it might be good to have `self.token` default to using `lookup('env', 'OC_TOKEN')` rather than supplying this value for every module invocation.